apis/v1alpha1/gateway_types.go

```diff
@@ -480,36 +480,41 @@ type RouteBindingSelector struct {
 	Resource string `json:"resource"`
 }
 
+// GatewaySelectType specifies where Routes should be selected by a Gateway.
+type GatewaySelectType string
+
+const (
+	// GatewaySelectFromSelector indicates that only Routes in namespaces
+	// selected by the selector may be used by this Gateway.
+	GatewaySelectFromSelector GatewaySelectType = "Selector"
+	// GatewaySelectSameNamespace indicates that Only Routes in the same
+	// namespace may be used by this Gateway.
+	GatewaySelectSameNamespace GatewaySelectType = "SameNamespace"
+)
+
 // RouteNamespaces indicate which namespaces Routes should be selected from.
 type RouteNamespaces struct {
-	// NamespaceSelector is a selector of namespaces that Routes should be
-	// selected from. This is a standard Kubernetes LabelSelector, a label query
-	// over a set of resources. The result of matchLabels and matchExpressions
-	// are ANDed. Controllers must not support Routes in namespaces outside this
-	// selector.
-	//
-	// An empty selector (default) indicates that Routes in any namespace can be
-	// selected.
-	//
-	// The OnlySameNamespace field takes precedence over this field. This
-	// selector will only take effect when OnlySameNamespace is false.
+	// From indicates where Routes will be selected for this Gateway. Possible
+	// values are:
+	// * Selector: Routes in namespaces selected by the selector may be used by
+	//   this Gateway.
+	// * SameNamespace: Only Routes in the same namespace may be used by this
+	//   Gateway.
 	//
 	// Support: Core
 	//
-	// +optional
-	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+	// +kubebuilder:validation:Enum=Selector;SameNamespace
+	// +kubebuilder:default=SameNamespace
+	From string `json:"from,omitempty"`
 
-	// OnlySameNamespace is a boolean used to indicate if Route references are
-	// limited to the same Namespace as the Gateway. When true, only Routes
-	// within the same Namespace as the Gateway should be selected.
-	//
-	// This field takes precedence over the NamespaceSelector field. That
-	// selector should only take effect when this field is set to false.
+	// Selector must be specified when From is set to "Selector". In that case,
+	// only Routes in Namespaces matching this Selector will be selected by this
+	// Gateway. This field is ignored for other values of "From".
```

Comment:
I think it's worth mentioning in the docs that an empty selector will select all ns's. I also like the idea of adding an `AllNamespaces` enum to make this even clearer to users.

Reply:
Yeah I think that does make sense. I added an `AllNamespaces` option here for simpler config.